### ISSUE TITLE
The DataTransfer interface is now exposed in Chromium and WebKit

### ISSUE THREAD

**Issue by foolip** [FineUploader/fine-uploader#1346]
https://github.com/FineUploader/fine-uploader/blob/master/client/js/features.js#L40 says "only way to test for Filesystem API support since webkit does not expose the DataTransfer interface" but the DataTransfer interface is now exposed.

Please consider first testing for DataTransfer and falling back to the existing code if browser usage of those old browsers is still high enough. The (qq.chrome() || qq.opera()) checks exclude Chromium-based browsers like Yandex and Coc Coc and will stop working if navigator.vendor is changed:
https://www.w3.org/Bugs/Public/show_bug.cgi?id=27786

**Comment by rnicholus** [comment]
The check you linked to is specific to folder-based DnD uploads.  I don't see how simply exposing this interface allows us to determine at load time if the browser supports DnD folder uploads.  It seems the only way to determine this is [by inspecting the `DataTransfer` instance associated with a drop event](https://github.com/FineUploader/fine-uploader/blob/5.1.3/client/js/util.js#L387).

**Comment by foolip** [comment]
OK, the "does not expose the DataTransfer interface" made me think exposing the DataTransfer interface would fix the problem. Can browsers that don't have webkitGetAsEntry just never drop folders, then?

**Comment by rnicholus** [comment]
The `webkitGetAsEntry` method on a `DataTransferItem` is a Blink-specific prefixed method to access the associated `Entry` for the dropped item.  The `Entry` interface is part of the Filesystem API, which used to be part of the HTML5 spec, but was recently dropped.  No other browser will ever support the Filesystem API or any associated interfaces, and it is certainly possible that Chrome will remove support in the future.  Handling a dropped folder in a drop event handler requires support of these interfaces.  Users of other browsers can certainly drop folders, but there is no way to parse them (or even determine that a folder has been dropped).  

This is a big gap in browser-based file support that only Chrome (and other browsers using Blink) have addressed.

**Comment by foolip** [comment]
Ah, so the problem is that the `DataTransferItem` interface isn't exposed. If it were, you could check for `DataTransferItem.prototype.webkitGetAsEntry`, right?

I'm going to fix this in Blink now: http://code.google.com/p/chromium/issues/detail?id=448696

**Comment by foolip** [comment]
The fix has landed, if you try Chrome Canary in a few days you should be able to check if my suggested replacement would work or not. You'd have to keep the sniffing as a fallback for a years or so, of course.

**Comment by rnicholus** [comment]
Awesome!  We'll have to decide when to remove the UA string parsing code here.

**Comment by foolip** [comment]
How about the "only way to test for complete Clipboard API support at this time" case, do you need something to detect support for that as well?

**Comment by rnicholus** [comment]
I can't remember at the moment.  I believe that the issue was the presence of the appropriate APIs in other browsers, but incomplete implementations.  I'll update once I get a chance to go over some history.

**Comment by rnicholus** [comment]
I think the issue is that Chrome allows a copied image (in the clipboard) to be pasted into any element, and other browsers require that element to be a contenteditable element.  There are some other issues as well, described in #774.  Not sure if this can be detected without UA string parsing.

**Comment by foolip** [comment]
Yeah, this does sound like something that's hard to do feature detection for. Would you happen to have a simple demo to illustrate the difference? I just want to make sure we're talking about the same thing, so that I can investigate what's correct per spec and maybe file bugs on the other browsers, if it hasn't already been reported of course.

**Comment by rnicholus** [comment]
The Clipboard API spec says [the following in the paste action section](http://www.w3.org/TR/clipboard-apis/#the-paste-action):

>If the cursor is in an editable element, the default action is to insert clipboard data in the most suitable format supported for the given context.

>The paste event has no default action in a non-editable context, but the event fires regardless.

Blink-based UAs fire paste on editable and non-editable elements.  Others only on editable elements.  So, it seems as if other browsers have not completely implemented the spec.
